### DISCUSSION TITLE
Is `recycle` method from factory random?

### DISCUSSION THREAD

**Discussion by Barbapapazes** [laravel/framework#53027] [Q&A]
Hello 👋,

There is something I'm not understanding with the `recycle` method.

```php
$emoji = Emoji::all();

$posts = Post::factory()
    ->count(50)
    ->create();

$users = User::factory()
    ->count(50)
    ->create();

$reactions = Reaction::factory()
    ->count(500)
    ->recycle([$posts, $emoji, $users])
    ->create();
```

With the following code, I expect that the first reaction to be post=1, emoji=1, user=1, then post=1, emoji=1, user=2 (or something like this but at the end, a predictable sequence where the recycle loop over small collections).

But, I've this in my database:

![Screenshot 2024-10-04 at 10 01 52](https://github.com/user-attachments/assets/adcddf8a-93f3-4597-8f4b-12569278f15f)

Which is clearly not a predictable sequence.

My issue is that I have the following SQL rule:

```php
$table->unique(['post_id', 'emoji_id', 'user_id']);
```

And the randomness cause SQL unique constraint to fail during seeding.

```
   Illuminate\Database\UniqueConstraintViolationException 

  SQLSTATE[23000]: Integrity constraint violation: 19 UNIQUE constraint failed: reactions.post_id, reactions.emoji_id, reactions.user_id (Connection: sqlite, SQL: insert into "reactions" ("post_id", "emoji_id", "user_id", "updated_at", "created_at") values (17, 4, 38, 2024-10-04 07:57:32, 2024-10-04 07:57:32))
```

How could I solve this issue?

**Comment by markunsworth** [comment]
Yes, it looks like the models are taken at random from the recycle collections in the factory.

```php
public function getRandomRecycledModel($modelClassName)
{
    return $this->recycle->get($modelClassName)?->random();
}

```

I think creating models where unique constraints are necessary is a fairly common use case, i tend to code around it in the seeders by generating a `Sequence` that I know is correct and passing that in but it feels like the factory should/could handle this. 

I'd propose that the `recycle` method was extended to allow something like a `useOnce` flag or perhaps a `recycleOnce` method was added for passing in models, so that when a model is used inside the factory it's popped off the recycle collection internally.
